clang/test/CodeGenHLSL/ArrayAssignable.hlsl

```diff
@@ -0,0 +1,50 @@
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.3-library -emit-llvm -disable-llvm-passes -o - %s | FileCheck %s --enable-var-scope
+
+// CHECK-LABEL: define void {{.*}}arr_assign1
+// CHECK: [[Arr:%.*]] = alloca [2 x i32], align 4
+// CHECK: [[Arr2:%.*]] = alloca [2 x i32], align 4
+// CHECK: [[Tmp:%.*]] = alloca [2 x i32], align 4
+// CHECK: call void @llvm.memcpy.p0.p0.i32(ptr align 4 [[Arr]], ptr align 4 {{@.*}}, i32 8, i1 false)
+// CHECK: call void @llvm.memset.p0.i32(ptr align 4 [[Arr2]], i8 0, i32 8, i1 false)
+// CHECK: call void @llvm.memcpy.p0.p0.i32(ptr align 4 [[Arr]], ptr align 4 [[Arr2]], i32 8, i1 false)
+// CHECK: call void @llvm.memcpy.p0.p0.i32(ptr align 4 [[Tmp]], ptr align 4 [[Arr]], i32 8, i1 false)
```

Comment:
Why are there 3 allocas here instead of 2? It looks to me like these memcpy/memsets are doing:
1. Copy the constant `{0, 1}` array into `Arr`
2. Zero initialize `Arr2`
3. Copy `Arr2` into `Arr`
4. Copy `Arr` into some other array?

I'm not sure why (4) is happening but it looks suspicious.

Reply:
I believe the Tmp exists so 'A = B =C' is possible.

Reply:
The problem is in how the codegen is emitting the lvalue. An example of the problem is this code:

```hlsl
export int fn() {
  int Arr[2] = {0, 1};
  int Arr2[2] = {1, 2};
  (Arr = Arr2)[0] = 6;
  return Arr[0] + Arr[1];
}
```
[DXC has this bug too](https://godbolt.org/z/M97qG9MfP).

Reply:
Is this a bug? I thought that was the correct behavior.

Reply:
> I believe the Tmp exists so 'A = B =C' is possible.

That doesn't seem correct. In the test case with `A = B = C` there is also a Tmp, but it's assigned after all three of A, B, and C

> The problem is in how the codegen is emitting the lvalue.

Ah, so the temporary value here is the (unused) lvalue for the result of the `A = B` expression?

Reply:
Yes, and that would enable Arr = Arr2 = Arr3 = Arr4 and so on.   Is it not an rvalue?

Reply:
The result of a builtin assignment operator is in fact an lvalue. This is described in [basic.lval](https://timsong-cpp.github.io/cppwp/n4659/basic.lval) and [expr.ass])https://timsong-cpp.github.io/cppwp/n4659/expr.ass) but is more clearly summarized on [cppreference's value category page](https://en.cppreference.com/w/cpp/language/value_category). When an assignment operator is used in an rvalue context it goes through [lvalue-to-rvalue conversion](https://timsong-cpp.github.io/cppwp/n4659/conv.lval)

Reply:
Ah, yeah. We do want to make sure that the assignment returns an lvalue and if we try to use it in an rvalue context it goes through an lvalue-to-rvalue conversion to do so. The PR as is seems to result in a jumble of C and C++ semantics, which is why it ends up doing these silly things.

It may be instructional to look at [an example with pointers](https://godbolt.org/z/6f3rq7r53) and [one with structs](
https://godbolt.org/z/8ex335r7W). Specifically, in the AST view you can see the LValueToRValue casts in these. Note that the struct case (the latter) is probably a bit closer semantically, but since it ends up using a call to `operator=` on the struct it's a bit harder to see what's going on. The pointer case (the former) is a little bit cleaner with respect to being able to read the AST output.